### PR TITLE
Usp 4288 implement start task command

### PR DESCRIPTION
**Pull request by agehrin50**

### CODE REVIEW

File: OPCUA/StartTaskCommandHandler.cs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    {
+        _logger = logger;
+        _opcuaService = opcuaService;
+    }
+
+    public string CommandName => "Start task";
+
+    public string GetResponseData(CallMethodResult result)
+    {
+        return (StatusCode.IsGood(result.StatusCode) ? 1 : 0).ToString();
```
Comment (Gishe) [inline OPCUA/StartTaskCommandHandler.cs:30]: You can do 
```csharp
return StatusCode.IsGood(result.StatusCode) ? "1" : "0";
```

File: OPCUA/GetTaskCommandHandler.cs [hunk 2]
```diff
@@ -10,28 +12,41 @@
...
+
+    public async Task<StringResponse> Handle(GetTasksRequest request, CancellationToken cancellationToken)
+    {
+        var printerId = request.data.ExtractPrinterId();
 
         var response = await _opcuaService.CallMethodAsync(printerId, OPCUAMethods.GetStoredMessageList.ToString(),
-            cancellationToken);
+            cancellationToken, Array.Empty<Variant>());
 
         if (response is null)
```
Comment (erikgabbard) [inline OPCUA/GetTaskCommandHandler.cs:51]:
```suggestion
        if (response is null || response.OutputArguments is null)
        {
            _logger.LogError("Get tasks OPCUA call failed");

            throw new OPCUACommunicationFailedException("Get tasks OPCUA call failed");
        }

        if (response.OutputArguments.Length != ExpectedOutputArgsLength)
        {
            _logger.LogError(
                "Output argument is not correct length. Expected {ExpectedOutputArgsLength}, got {outputArgumentsLength}",
                ExpectedOutputArgsLength, response.OutputArguments.Length);

            throw new InvalidDataException("Output argument is not correct length.");
        }

        return new StringResponse(CommandName, printerId, GetResponseData(response));
```

File: OPCUA/ResmarkOPCUAService.cs [hunk 3]
```diff
@@ -42,7 +42,8 @@ public ResmarkOPCUAService(IConfiguration configuration, ILogger<ResmarkOPCUASer
     public async Task<CallMethodResult?> CallMethodAsync(
         string printerId,
         string method,
-        CancellationToken stoppingToken)
+        CancellationToken stoppingToken,
```
Comment (erikgabbard) [inline OPCUA/ResmarkOPCUAService.cs:45]: See previous comment about arguments

File: OPCUA/ResmarkOPCUAService.cs [hunk 4]
```diff
@@ -62,14 +63,15 @@ public ResmarkOPCUAService(IConfiguration configuration, ILogger<ResmarkOPCUASer
 
     private async Task<CallResponse> MakeCallRequest(
         string method,
+        Variant[]? inputArgs,
         CancellationToken stoppingToken,
```
Comment (erikgabbard) [inline OPCUA/ResmarkOPCUAService.cs:67]: See previous comment about arguments

File: OPCUA/ResmarkOPCUAService.cs [hunk 5]
```diff
@@ -62,14 +63,15 @@ public ResmarkOPCUAService(IConfiguration configuration, ILogger<ResmarkOPCUASer
 
     private async Task<CallResponse> MakeCallRequest(
         string method,
+        Variant[]? inputArgs,
```
Comment (Drise13) [inline OPCUA/ResmarkOPCUAService.cs:66]: There should be a seperate overload of this function that handles input args. Don't force all consumers to pass in empties.

File: StringResponse.cs [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+public class StringResponse
+{
+    public StringResponse(string commandName, string printerId, string responseData)
+    {
+        Data = FormatResponse(commandName, printerId, responseData);
+    }
+
+    public string Data { get; set; }
+
+    private string FormatResponse(string commandName, string? printerId, string responseData)
```
Comment (Drise13) [inline StringResponse.cs:14]: Why is the id nullable? Why is it not nullable in the caller? Pick one and stick with it (likely not nullable)

File: TCP/BoxwriterTCPHandler.cs [hunk 7]
```diff
@@ -15,12 +15,13 @@ public BoxwriterTCPHandler(IMediator mediator)
 
     public async Task<StringResponse> Handle(TCPRequest request, CancellationToken cancellationToken)
     {
-        var tcpRequest = request.data switch
+        object tcpRequest = request.data switch
```
Comment (erikgabbard) [inline TCP/BoxwriterTCPHandler.cs:18]:
```suggestion
        return request.data switch
        {
            var data when data.StartsWith("Get") => 
                await _mediator
                    .Send(new GetTasksRequest(data))
                    .ConfigureAwait(false),
            var data when data.StartsWith("Start") => 
                await _mediator
                    .Send(new StartTaskRequest(data))
                    .ConfigureAwait(false),
            _ => throw new InvalidOperationException(),
        };
```

Comment (Gishe) [inline TCP/BoxwriterTCPHandler.cs:18]: If you define this like this
```csharp
 IRequest<StringResponse> tcpRequest = request.data switch
```
You don't need to cast it and you lose the nullable warnings.

**Review by Gishe — CHANGES_REQUESTED** [review]

File: StringResponse.cs
Comment (erikgabbard) [inline StringResponse.cs:1]: Consider moving this class to .\Responses\StringResponse.cs making the namespace `BoxwriterResmarkInterop.Responses.StringResponse`.

File: OPCUA/StartTaskCommandHandler.cs [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    public async Task<StringResponse> Handle(StartTaskRequest request, CancellationToken cancellationToken)
+    {
+        var printerId = request.data.ExtractPrinterId();
+        var messageName = request.data.ExtractMessageName();
+
+        var inputArgs = new[] { new Variant(messageName) };
+
+        var response = await _opcuaService.CallMethodAsync(printerId, OPCUAMethods.PrintStoredMessage.ToString(),
```
[the remark this answers is not in the document]
Comment (Gishe) [reply OPCUA/StartTaskCommandHandler.cs:40]: Oh I guess I miss understood the sequence.  We want load task to select and start task to start printing I guess.  Its hard when the system's don't work 100% the same.

File: StringResponse.cs (hunk 6 above)
Comment (erikgabbard) [inline StringResponse.cs:14]:
```suggestion
    private string FormatResponse(string commandName, string? printerId, string responseData) => 
        $"{StartToken}{string.Join(TokenSeparator, commandName, printerId, responseData)}{EndToken}";
```

**Review by erikgabbard — CHANGES_REQUESTED** [review]

File: TCP/BoxwriterTCPHandler.cs (hunk 7 above)
Comment (Drise13) [reply TCP/BoxwriterTCPHandler.cs:18]:
```suggestion

        return await _mediator
                    .Send(request.data switch
        {
            var data when data.StartsWith("Get") => new GetTasksRequest(data),
            var data when data.StartsWith("Start") => new StartTaskRequest(data))
            _ => throw new InvalidOperationException(),
        }.ConfigureAwait(false);

```

Comment (Gishe) [reply TCP/BoxwriterTCPHandler.cs:18]: I disagree with both of these.  Getting the `IRequest<StringResponse>` from the switch is much easier to read and edit.

File: OPCUA/StartTaskCommandHandler.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (agehrin50) [reply OPCUA/StartTaskCommandHandler.cs]: I updated the ticket to reflect our conversation. Start task is essentially resume printing

**Review by Gishe — APPROVED** [review]

File: OPCUA/GetTaskCommandHandler.cs [hunk 9]
```diff
@@ -10,28 +12,41 @@
...
 
-    public async Task<StringResponse> Handle(GetTaskRequest request, CancellationToken cancellationToken)
+    public string GetResponseData(CallMethodResult result)
     {
-        var printerId = ExtractPrinterId(request.data);
+        var outputArguments = result.OutputArguments;
+
+        if (outputArguments?[1].Value is string[] args)
+        {
+            return args.Aggregate((current, arg) => current + TokenSeparator + arg);
```
[the remark this answers is not in the document]
Comment (agehrin50) [reply OPCUA/GetTaskCommandHandler.cs]: Erik and I found that this was not a simple string.join that could be done here. In the spirit of finishing the MVP, we will be leaving it as is.

File: Extensions/StringExtensions.cs [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+    {
+        data = data.Trim(StartToken);
+        data = data.Trim(EndToken);
+
+        var printerId = data.Split(TokenSeparator)[1];
+
+        return printerId;
+    }
+
+    public static string ExtractMessageName(this string data)
```
[the remark this answers is not in the document]
Comment (agehrin50) [reply Extensions/StringExtensions.cs:17]: Reworked

File: Handlers/StartTaskCommandHandler.cs [hunk 11]
```diff
@@ -1,18 +1,23 @@
...
+using OPCUA;
+
+using Requests;
+
 using Workstation.ServiceModel.Ua;
 
 public class StartTaskCommandHandler : IRequestHandler<StartTaskRequest, StringResponse>
 {
     private const string CommandName = "Start task";
+    private const int TaskNumber = 1;
```
Comment (erikgabbard) [inline Handlers/StartTaskCommandHandler.cs:20]: Should this go with the other constants?

File: Handlers/StartTaskCommandHandler.cs [hunk 12]
```diff
@@ -22,18 +27,12 @@ public StartTaskCommandHandler(ILogger<StartTaskCommandHandler> logger, IOPCUASe
...
     {
         var printerId = request.Data.ExtractPrinterId();
-        var messageName = request.Data.ExtractMessageName();
-
-        var inputArgs = new[] { new Variant(messageName) };
 
-        var response = await _opcuaService.CallMethodAsync(printerId, OPCUAMethods.PrintStoredMessage.ToString(),
-            cancellationToken, inputArgs);
+        var response =
+            await _opcuaService.CallMethodAsync(printerId, OPCUAMethods.ResumePrinting.ToString(), cancellationToken, TaskNumber);
```
Comment (erikgabbard) [inline Handlers/StartTaskCommandHandler.cs:35]: `ConfigureAwait(false)`

File: OPCUA/ResmarkOPCUAService.cs [hunk 13]
```diff
@@ -54,24 +54,42 @@ public ResmarkOPCUAService(IConfiguration configuration, ILogger<ResmarkOPCUASer
 
         var channel = await OpenChannel(printerId, stoppingToken);
 
-        var response = await MakeCallRequest(method, inputArgs, stoppingToken, channel);
+        var response = await MakeCallRequest(method, channel, stoppingToken, inputArgs);
```
Comment (erikgabbard) [inline OPCUA/ResmarkOPCUAService.cs:57]: `ConfigureAwait(false)`

File: OPCUA/ResmarkOPCUAService.cs [hunk 14]
```diff
@@ -54,24 +54,42 @@ public ResmarkOPCUAService(IConfiguration configuration, ILogger<ResmarkOPCUASer
...
+        return await CallMethodAsync(printerId, method, stoppingToken, new Variant[] { taskNumber });
+    }
+
+    public async Task<CallMethodResult?> CallMethodAsync(
+        string printerId,
+        string method,
+        CancellationToken stoppingToken,
+        string[] inputArgs)
+    {
+        return await CallMethodAsync(printerId, method, stoppingToken, inputArgs.ToVariantArray());
```
Comment (erikgabbard) [inline OPCUA/ResmarkOPCUAService.cs:79]: `ConfigureAwait(false)`

File: Handlers/GetTasksRequestHandler.cs [hunk 15]
```diff
@@ -28,21 +31,17 @@ public GetTasksRequestHandler(IOPCUAService opcuaService, ILogger<GetTasksReques
...
+    private static IEnumerable<string> GetResponseData(CallMethodResult result) =>
+        result.OutputArguments?[1].Value as string[] ?? new [] { Constants.NoMessages };
 
     public async Task<StringResponse> Handle(GetTasksRequest request, CancellationToken cancellationToken)
     {
         var printerId = request.Data.ExtractPrinterId();
 
         var response = await _opcuaService.CallMethodAsync(printerId, OPCUAMethods.GetStoredMessageList.ToString(),
-            cancellationToken, Array.Empty<Variant>());
+            cancellationToken);
```
Comment (erikgabbard) [inline Handlers/GetTasksRequestHandler.cs:42]: `ConfigureAwait(false)`

**Review by erikgabbard — CHANGES_REQUESTED** [review]
Be sure to remember `ConfigureAwait(false)` when awaiting.

File: Handlers/StartTaskCommandHandler.cs (hunk 11 above)
Comment (agehrin50) [reply Handlers/StartTaskCommandHandler.cs:20]: I would agree

**Review by erikgabbard — APPROVED** [review]

**Review by Drise13 — APPROVED** [review]

**Pull request merged**